### PR TITLE
add web server to e2et

### PR DESCRIPTION
**Pull request by GeorgeJahad**
# Resolves

https://jira.rax.io/browse/SALUS-697
https://jira.rax.io/browse/SALUS-667

# What

Added features to facilite use in production, the main one being a webserver.

By starting with the *--web-server-port=xxx* command line parameter, the program starts a webserver at the corresponding port.  An http get of that port will run the e2et and return 200 if it succeeds or 500, along with the log messages if it fails.

The other changes are:
* the config vars were rearranged a bit to better support env var overrides.
* kills any envoy instances from still working from previous runs, so they don't interfere
* downloads the envoy exe, and gives it a unique name, rather than expecting it in the gopath
* fixed the identity url config, so it would support staging identity

### CODE REVIEW

File: e2et/main.go [hunk 1]
```diff
@@ -44,53 +48,49 @@ import (
...
 	c.currentUUID = uuid.NewV4()
 	c.id = strings.Replace(c.currentUUID.String(), "-", "", -1)
 	c.privateZoneId = "privateZone_" + c.id
 	c.resourceId = "resourceId_" + c.id
-	c.tenantId = viper.GetString("tenantId")
-	c.publicApiUrl = viper.GetString("publicApiUrl")
-	c.adminApiUrl = viper.GetString("adminApiUrl")
+	c.tenantId = viper.GetString("tenant.id")
+	c.publicApiUrl = viper.GetString("public.api.url")
+	c.adminApiUrl = viper.GetString("admin.api.url")
```
Comment (itzg) [inline e2et/main.go]: If you weren't so far in with viper (like we are also with the Envoy code) I'd suggest you would find [flagsfiller](https://pkg.go.dev/github.com/itzg/go-flagsfiller?tab=doc) to be much less tedious. It doesn't directly support reading from config files, but can always be augmented with a yaml decoding into the config object.

File: e2et/main.go [hunk 2]
```diff
@@ -125,7 +139,7 @@ func main() {
 	select {
 	case <-eventFound:
 		log.Println("events returned from kafka successfully")
-	case <-time.After(10 * time.Minute):
+	case <-time.After(5 * time.Minute):
```
Comment (itzg) [inline e2et/main.go]: Maybe this should be configurable.

File: e2et/main.go [hunk 3]
```diff
@@ -151,11 +165,28 @@ func initEnvoy(c config, releaseId string) (cmd *exec.Cmd) {
...
-		c.certDir, c.regularApiKey, c.regularId, c.authUrl, c.ambassadorAddress})
+		c.certDir, c.regularApiKey, c.regularId, c.authUrl, c.ambassadorAddress,
+		c.identityUrl})
 	checkErr(err, "creating envoy template")
 	err = f.Close()
 	checkErr(err, "closing envoy config file: "+configFileName)
-	cmd = exec.Command(os.Getenv("GOPATH")+"/bin/telemetry-envoy", "run", "--config="+configFileName)
+	var tarballURL string
+	if runtime.GOOS == "darwin" {
+		tarballURL = "https://github.com/racker/salus-telemetry-envoy/releases/download/0.13.0/telemetry-envoy_0.13.0_Darwin_x86_64.tar.gz"
```
Comment (itzg) [inline e2et/main.go]: Version should be configurable

File: e2et/main.go [hunk 4]
```diff
@@ -151,11 +165,28 @@ func initEnvoy(c config, releaseId string) (cmd *exec.Cmd) {
...
+	} else {
+		tarballURL = "https://github.com/racker/salus-telemetry-envoy/releases/download/0.13.0/telemetry-envoy_0.13.0_Linux_x86_64.tar.gz"
+	}
+	cmd = exec.Command("curl", "-L", "-o", c.dir+"/envoy.tar.gz", tarballURL)
+	err = cmd.Run()
+	checkErr(err, "downloading tar")
+	cmd = exec.Command("tar", "-xf", c.dir+"/envoy.tar.gz", "-C", c.dir)
+	err = cmd.Run()
+	checkErr(err, "decompressing tarball")
```
Comment (itzg) [inline e2et/main.go]: In theory you can do this natively, but it's certainly a lot more tedious and lines of code than the curl/tag usage:

https://github.com/racker/salus-telemetry-envoy/blob/1e502d6c821350c99b721812317c66e7de54f0cf/agents/agents.go#L97

File: e2et/types.go [hunk 5]
```diff
@@ -20,6 +20,7 @@ package main
 
 import (
 	"github.com/satori/go.uuid"
+	"golang.org/x/sync/semaphore"
```
Comment (itzg) [inline e2et/types.go]: Oh yeah, Go's `sync` package doesn't include a semaphore...I guess they expect one to be built from a `Cond`.

File: e2et/main.go (hunk 4 above)
Comment (GeorgeJahad) [reply e2et/main.go]: nice!  are you ok with curl and tag, or should i change it?

File: e2et/types.go (hunk 5 above)
Comment (GeorgeJahad) [reply e2et/types.go]: yeh i started building something and then saw this.

File: e2et/main.go (hunk 2 above)
Comment (GeorgeJahad) [reply e2et/main.go]: done

File: e2et/main.go (hunk 3 above)
Comment (GeorgeJahad) [reply e2et/main.go]: done

File: e2et/main.go (hunk 1 above)
Comment (GeorgeJahad) [reply e2et/main.go]: it's fine with me now.  i may change it in the future.

File: e2et/config-local.yml [hunk 6]
```diff
@@ -1,8 +1,11 @@
...
+tenant.id: aaaaaa
+public.api.url: http://localhost:8080/
+admin.api.url: http://localhost:8888/
+cert.dir: ../../dev/certs
+kafka.brokers:
   - localhost:9092
-eventTopic: salus.events.json
+event.topic: salus.events.json
+event.timeout.minutes: 5
+envoy.tarball.linux: https://github.com/racker/salus-telemetry-envoy/releases/download/0.13.0/telemetry-envoy_0.13.0_Linux_x86_64.tar.gz
```
Comment (jjbuchan) [inline e2et/config-local.yml]: Could also hardcode `latest` for these things.

Add a grep to this for each os `curl -sL https://api.github.com/repos/racker/salus-telemetry-envoy/releases/latest | jq -r '.assets[].browser_download_url'`

File: e2et/main.go (hunk 4 above)
Comment (itzg) [reply e2et/main.go]: Yeah, using curl and tar work for me. The only advantage to native code is if you wanted to created a stripped down docker image without those included...so perhaps a later option to ponder.

File: e2et/config-local.yml (hunk 6 above)
Comment (itzg) [reply e2et/config-local.yml]: Interesting idea, but I wonder if you can use shell piping within the exec call. Could always emulate the jq aspect with native JSON parsing of the releases query.

Given the nature of the tool though I'd recommend just sticking with a configured URL.

File: e2et/main.go [hunk 7]
```diff
@@ -91,6 +92,10 @@ func initConfig() config {
 	c.regularToken = getToken(c, c.regularId, c.regularApiKey, "")
 	c.adminToken = getToken(c, c.adminId, c.adminApiKey, c.adminPassword)
 	c.envoyExeName = "e2et-envoy-" + c.tenantId
+	c.envoyTarballDarwin = viper.GetString("envoy.tarball.darwin")
+	c.envoyTarballLinux = viper.GetString("envoy.tarball.linux")
+	c.envoyTimeout, err = strconv.Atoi(viper.GetString("envoy.timeout"))
+	checkErr(err, "error converting timeout: "+viper.GetString("envoy.timeout"))
```
Comment (itzg) [inline e2et/main.go]: Viper has support for durations and that would avoid the need to convert/check and then avoid requiring a particular unit.

https://godoc.org/github.com/spf13/viper#GetDuration

BTW, it should have actually been `event.timeout.minutes` as currently implemented.

Comment (GeorgeJahad) [reply e2et/main.go]: good point, thanks! fixed now.

File: e2et/config-local.yml (hunk 6 above)
Comment (jjbuchan) [reply e2et/config-local.yml]: > Given the nature of the tool 

Can you expand on what you mean there?

I was kinda thinking we still might want the ability to override what's used but have the default be to use `latest`, otherwise I could see us missing envoy bugs that might have otherwise been caught.  Removing a manual step from any envoy release seems beneficial.

No strong opinions from me but I'm curious what your reasoning is.

@GeorgeJahad this discussion isn't a blocker in any way.  This can easily be changed later if we want.

Comment (itzg) [reply e2et/config-local.yml]: For the first aspect, I was pointing out that piping curl into jq may not work with

https://golang.org/pkg/os/exec/#Command

As for the "nature of the tool", I just meant since this is not a mainline part of the product we itself, we have more liberty to just keep it simple since it isn't user facing.

Comment (GeorgeJahad) [reply e2et/config-local.yml]: To clarify my thinking, I haven't seen the perf test framework as a good way to test the envoys.  Envoys are being used in a fairly artificial way, (perf test mode,) just to generate load.  Given that, I haven't been so concerned about being sure the latest version of the envoy is being used, (as long as it supports perf test mode).  Maybe that's too cavalier of me, but that has been my thinking up to now.

File: e2et/templates.go [hunk 8]
```diff
@@ -113,13 +114,13 @@ const httpMonitorData = `{
...
   },
-  "interval": "PT30S",
+  "interval": "PT60S",
   "details": {
     "type": "remote",
     "monitoringZones": ["%s"],
     "plugin": {
       "type": "http_response",
-      "address": "http://localhost:%s",
+      "address": "http://l:80",
```
Comment (jjbuchan) [inline e2et/templates.go]: why the change?

File: e2et/config-local.yml [hunk 9]
```diff
@@ -1,8 +1,11 @@
...
-tenantId: aaaaaa
-publicApiUrl: http://localhost:8080/
-adminApiUrl: http://localhost:8888/
-certDir: ../../dev/certs
-kafkaBrokers:
+tenant.id: aaaaaa
+public.api.url: http://localhost:8080/
+admin.api.url: http://localhost:8888/
+cert.dir: ../../dev/certs
+kafka.brokers:
```
[the remark this answers is not in the document]
Comment (GeorgeJahad) [reply e2et/config-local.yml]: yes.

File: e2et/main.go [hunk 10]
```diff
@@ -493,3 +528,51 @@ func checkPresenceMonitor(c config) {
...
+		wr.WriteHeader(http.StatusLocked)
+		_, _ = wr.Write([]byte("e2et already running.\n"))
+		return
+	}
+	defer w.sem.Release(1)
+	log.Info("starting request")
+	buf := new(bytes.Buffer)
+	cmd := exec.Command(os.Args[0], "--config="+*w.cfgFileName)
+	cmd.Stdout = buf
+	cmd.Stderr = buf
```
[the remark this answers is not in the document]
Comment (GeorgeJahad) [reply e2et/main.go]: no, my way interleaves standard err and standard out.  your's keeps them separate.

File: e2et/templates.go (hunk 8 above)
Comment (GeorgeJahad) [reply e2et/templates.go]: it's not that important, but i wanted to get a unresolved host error at some point, and then forgot to change it back

File: e2et/main.go [hunk 11]
```diff
@@ -493,3 +528,51 @@ func checkPresenceMonitor(c config) {
...
+	log.Info("started e2et webServer")
+	err = http.Serve(listener, serverMux)
+	log.Fatalf("e2et server error %v", err)
+}
+
+func (w *webServer) handler(wr http.ResponseWriter, r *http.Request) {
+	if !w.sem.TryAcquire(1) {
+		log.Error("e2et already running")
+		wr.WriteHeader(http.StatusLocked)
+		_, _ = wr.Write([]byte("e2et already running.\n"))
```
[the remark this answers is not in the document]
Comment (GeorgeJahad) [reply e2et/main.go]: i think it keeps intellij from complaining about an unhandled error.

File: e2et/main.go [hunk 12]
```diff
@@ -493,3 +528,51 @@ func checkPresenceMonitor(c config) {
...
+
+func deleteEnvoyProcesses(c config) {
+	processes, err := process.Processes()
+	checkErr(err, "getting processes")
+	for _, p := range processes {
+		name, _ := p.Name()
+		if strings.Contains(name, c.envoyExeName) {
+			c, _ := p.Cmdline()
+			log.Info("killing envoy: " + c)
+			p.Kill()
```
[the remark this answers is not in the document]
Comment (GeorgeJahad) [reply e2et/main.go]: i'm only using "c" for the log message, so if it gets an error, the log message will be incomplete, but I'm ok with that.

It probably blocks until the kill signal is created.  But I'm not sure it blocks until  the process is actually killed.  I haven't seen that be a problem up to now.  I'll handle it if it becomes a problem.
